src/glfw/window.cc

```diff
@@ -0,0 +1,233 @@
+{
+    Window* window = getHandle( glfwWindow );
+
+    glfwDestroyWindow( glfwWindow );
+
+    logGLFWaction(
+        "destroy window",
+        getHandleAddressString( window ) //
+    );
+} // destroyWindow
+
+void
+Window::framebufferSizeCallback(
+    WindowType* glfwWindow,
+    int width,
+    int height //
+)
+{
+    Window* window = getHandle( glfwWindow );
+
+    logGLFWaction(
+        "framebuffer resize",
+        getHandleAddressString( window ),
+        "width = " + std::to_string( width ) + ", height = " + std::to_string( height ) //
+    );
+
+    window->m_resize_callback( window, width, height );
+} // framebufferSizeCallback
+
+FramebufferSizeType
+Window::getFramebufferSize() const&
+{
+    int width = 0;
+    int height = 0;
+
+    glfwGetFramebufferSize( m_handle.get(), &width, &height );
+
+    return std::make_pair( width, height );
+} // getFramebufferSize
+
+void
+Window::setWindowed() const&
+{
+    constexpr int xpos = 0;
+    constexpr int ypos = 0;
+
+    glfwSetWindowMonitor(
+        m_handle.get(),
+        nullptr,
+        xpos,
+        ypos,
+        getFramebufferWidth(),
+        getFramebufferHeight(),
+        GLFW_DONT_CARE );
+
+    logGLFWaction(
+        "choose windowed mode",
+        getHandleAddressString( this ) //
+    );
+} // setWindowed
+
+void
+Window::setFullscreen() const&
+{
+    GLFWmonitor* monitor = glfwGetPrimaryMonitor();
+
+    glfwSetWindowMonitor(
+        m_handle.get(),
+        monitor,
+        0,
+        0,
+        getFramebufferWidth(),
+        getFramebufferHeight(),
+        GLFW_DONT_CARE );
+
+    logGLFWaction( "choose fullscreen mode", getHandleAddressString( this ) );
+} // setFullscreen
+
+WindowManager::WindowManager( ErrorCallbackSignature* error_callback )
+{
+    glfwSetErrorCallback( error_callback );
+    glfwInit();
+
+    logGLFWaction( "initialize library" );
+} // WindowManager
+
+WindowManager::~WindowManager()
+{
+    glfwTerminate();
+
+    logGLFWaction( "terminate library" );
+} // ~WindowManager
+
+bool
+WindowManager::isSuitableVersion()
+{
+    int major = 0;
+    int minor = 0;
+
+    glfwGetVersion( &major, &minor, nullptr );
+
+    if ( major > k_min_major )
+    {
+        return true;
+    }
+
+    if ( major == k_min_major && minor >= k_min_minor )
+    {
+        return true;
+    }
+
+    return false;
+} // isSuitableVersion
+
+void
+WindowManager::initialize( ErrorCallbackSignature* error_callback )
+{
+    static WindowManager instance( error_callback );
+
+    if ( !isSuitableVersion() )
+    {
+        throw Error(
+            Error::k_user_error,
+            "GLFW minimal version: " + getMinVersionString() + "." //
+        );
+    }
+} // initialize
+
+std::span<std::string_view>
+WindowManager::getRequiredExtensions()
+{
+    uint32_t size = 0;
+    const char** extensions = glfwGetRequiredInstanceExtensions( &size );
+
+    static std::vector<std::string_view> vectorized{ extensions, extensions + size };
```

Comment:
Why is this vector static? Moreover, this can be be a std::span of std::string_view. Then there would be no reason to create a dynamically allocated vector with static storage duration.

Reply:
I can see that you were afraid to return a view to a local object and because of this you didn't write this function properly. 
An array return by ```glfwGetRequiredInstanceExtensions``` is stored somewhere in memory where GLFW manages it. Thus it is safe to return a view of the data.

```cpp
    const char** extensions = glfwGetRequiredInstanceExtensions( &size );
    return std::span<const char *>{extensions, extensions + size};
```

Reply:
Agree.